### PR TITLE
Typescript refactor: stub rxjs

### PR DESCRIPTION
**Pull request by bodograumann**
As discussed in #802 I have replaced RxJS with a stub when it is not really needed.
There is a new option for the typescript generator `useRxJS` which defaults to `false`.

Additionally I fixed some minor issues.

I noticed that the tests in `samples/client/petstore/typescript/tests` are not automatically generated, but statically committed. So that is another open todo?

CC @TiFu

### CODE REVIEW

File: modules/openapi-generator/src/main/resources/typescript/api/middleware.mustache [hunk 1]
```diff
@@ -55,12 +55,12 @@ export interface PromiseMiddleware {
 	 * @returns an observable of the updated request context
 	 *
 	 */
-    pre(context: RequestContext): Observable<RequestContext>;
```
Comment (TiFu) [inline modules/openapi-generator/src/main/resources/typescript/api/middleware.mustache]: Why was this changed to `Promise`?

**Comment by bodograumann** [comment]
### Compilation errors
I compiled locally. The `mvn` and generating the samples worked fine.
The tests did not work. I am not sure why that is. It was indeed confusing because the last time it worked, but when I checked out the typescript-refactor branch and rebuild everything, they still would not work. I guess I need to take a little more time to investigate this issue.

### Readme
Yeah, I did not run ensure-up-to-date.sh again. I was hoping it wasn't needed because last time it changed unrelated things differently than the CI.
Have added the docs line for the useRxJS switch manually now.

### Middleware
I changed it to what I think the original intention was. This is the only way that makes sense to me. Previously both interfaces `Middleware` and `PromiseMiddleware` where identical.
When thinking about how middlewares should work with RxJS stubed, I came to the conclusion, that only `PromiseMiddleware` which is then wrapped, can be used. The user should not be aware of the `Observable` stub, so he must work only with `Promise`.

**Comment by bodograumann** [comment]
Ok, got the tests to run on the original branch and found the problem:
We require the node types, but only implicitly depended on them via `@types/form-data`.
I have now added them as an explicit dependency.
Furthermore I have adjusted the dependencies in the tests according to the changes to the template and samples.

For the jquery tests to run I had to install webpack.

The tests where using `Observable.subscribe`. Afaics this is currently not really necessary, so I replaced it with the stubbed `Observable.toPromise().then`. If in the future it is needed, we can of course add it to the stub.

Not sure whether the jquery tests ran properly on my local machine, because I got 
> \> node ./dist/test-runner.js
>
> [error] Failed to load resource: the server responded with a status of 404 (Not Found)

But let's see what the ci says.

**Comment by bodograumann** [comment]
The failures don’t seem to be related to my changes.
What do you think, @TiFu?

**Comment by TiFu** [comment]
Thanks a lot for investigating this issue and the PR! 

- CircleCI is unrelated
- DroneIO as well
Going to close and reopen this PR to rerun the tests. If it fails again, i'll just merge anyway. 

> Compilation errors
> Readme
Sounds good

> Middleware
Oh the change was in `PromiseMiddleware`. Good catch.

**Comment by bodograumann** [comment]
Ok, forgot to remove `subscribe` from the jquery tests. Now it really is complete :-D
Tests seem good. Failure because of some OCAML stuff. Different from before but still unrelated to these changes.

**Comment by TiFu** [comment]
Great! Again, thanks for the PR which has been merged into ts-refactor.

**Comment by wing328** [comment]
@bodograumann thanks for the PR, which has been included in the v4.2.2 release: https://twitter.com/oas_generator/status/1201432648544972800
